py-nominal-streaming/python/nominal_streaming/nominal_dataset_stream.py

```diff
@@ -0,0 +1,311 @@
+"""Python-facing API for the streaming client.
+
+Example:
+-------
+import pathlib
+from datetime import datetime, timedelta, timezone
+
+from nominal_streaming import NominalStreamOpts, NominalDatasetStream
+
+# NOTE: may also use NominalStreamOpts.default() for sensible defaults that may be customized
+opts = NominalStreamOpts(
+    max_points_per_batch=250_000,
+    max_request_delay=timedelta(seconds=0.1),
+    max_buffered_requests=4,
+    num_upload_workers=8,
+    base_api_url="https://api.gov.nominal.io/api",
+    runtime_workers=8,
+)
+
+with (
+    NominalDatasetStream("api_key", opts)
+    .with_core_consumer("ri.catalog.dataset...")
+    .with_file_fallback(pathlib.Path("/tmp/fallback.avro")) as stream
+):
+    stream.enqueue("chanA", datetime.now(timezone.utc), 1.23, tags={"site": "a1"})
+    stream.enqueue_batch("chanB", [0, 1_000_000_000], [5, 6], tags={"phase": "prod"})
+    stream.enqueue_from_dict(0, {"chanC": "ok", "chanD": 7}, tags={"who": "tester"})
+
+"""
+
+from __future__ import annotations
+
+import datetime
+import logging
+import pathlib
+import signal
+from types import TracebackType
+from typing import Mapping, Sequence, Type
+
+import dateutil
+
+from nominal_streaming._nominal_streaming import (
+    NominalStreamOpts,
+    _NominalDatasetStream,
+)
+
+logger = logging.getLogger(__name__)
+
+TimestampLike = str | int | datetime.datetime
+DataType = int | float | str
+
+
+def _parse_timestamp(ts: str | int | datetime.datetime) -> int:
+    if isinstance(ts, int):
+        return ts
+    elif isinstance(ts, datetime.datetime):
+        secs = ts.astimezone(datetime.timezone.utc).timestamp()
+        return int(secs * 1e9)
+    else:
+        # TODO(drake): by involving dateutil, this chops off any nano level precision provided
+        #              in the timestamp. Update to not lose precision when converting to absolute nanos.
+        secs = dateutil.parser.parse(ts).astimezone(datetime.timezone.utc).timestamp()
+        return int(secs * 1e9)
+
+
+class NominalDatasetStream:
+    """Top-level python wrapper for the Rust streaming client to Nominal."""
+
+    def __init__(self, auth_header: str, opts: NominalStreamOpts):
+        """Initializer for dataset stream.
+
+        Args:
+            auth_header: API key or access token to the Nominal API
+            opts: Optional options for the underlying stream
+        """
+        self._auth_header = auth_header
+        self._opts = opts
+        self._impl = _NominalDatasetStream(self._opts)
+        self._old_sigint = None
+
+    @classmethod
+    def create(
+        cls,
+        auth_header: str,
+        max_points_per_batch: int | None = None,
+        max_request_delay_secs: float | None = None,
+        max_buffered_requests: int | None = None,
+        num_upload_workers: int | None = None,
+        num_runtime_workers: int | None = None,
+        api_base_url: str | None = None,
```

Comment:
Is there ever a use-case where the API base URL isn't provided by the caller? I.e. should this just be required?